### PR TITLE
Update Neo4j extension to use Spring Data Neo4j.

### PR DESCRIPTION
**Pull request by meistermeier**
This is a rather large PR because it will remove the Neo4j-OGM dependency and replace it with the Spring Data Neo4j.

Some background information to clarify the situation:
The official Spring Data Neo4j was rewritten a few years ago to not depend on Neo4j-OGM anymore (in Spring Boot version since 2.4).
This current (old) implementation is based on the fact that the Spring Boot starter still brings Neo4j-OGM as a dependency and makes use of the configured `SessionFactory` instance, that does not exist anymore.

Pros: No need to specify the query parts anymore but only one query/statement based on the Neo4j Cypher-DSL (https://github.com/neo4j-contrib/cypher-dsl) for readers
Cons: The writer needs now, instead of the single `SessionFactory`, 
* the `Neo4jMappingContext` to have knowledge about the entities
* the `Neo4jTemplate` for loading and mapping entities
* the `(Neo4j)Driver` for the simple delete operations.

While changing the mapping layer component, I decided also to upgrade all other dependencies to latest.

### CODE REVIEW

File: spring-batch-neo4j/README.md [hunk 1]
```diff
@@ -7,25 +7,100 @@ This extension contains an `ItemReader` and `ItemWriter` implementations for [Ne
...
+Neo4jItemWriter<User> writer = new Neo4jItemWriterBuilder<User>()
+        .neo4jTemplate(neo4jTemplate)
+        .neo4jDriver(driver)
+        .neo4jMappingContext(mappingContext)
         .build();
+```
+
+## Minimal Spring Boot example
+
+With a Spring Boot application containing the additional dependencies `spring-boot-starter-neo4j` and `spring-batch-neo4j`,
```
Comment (michael-simons) [inline spring-batch-neo4j/README.md:31]: That should go together with the following sentence.

x, y, z are needed and b must be excluded. See the following gradle for a minimal example,

**Comment by mbollemeijer** [comment]
Is there any ETA on when this will be merged ?

**Review by michael-simons — APPROVED** [review]

**Comment by meistermeier** [comment]
If there is demand to get this in, I would like to get some feedback from @mdeinum or @dgray16 here (at least they seem pretty active in the past).
No problem for me to check possible upgrades and apply some needed adjustments before merge.

If this is up to us alone, I am technical able to do the merge, but please let us know.

File: spring-batch-neo4j/src/main/java/org/springframework/batch/extensions/neo4j/builder/Neo4jItemWriterBuilder.java [hunk 2]
```diff
@@ -49,14 +54,36 @@ public Neo4jItemWriterBuilder<T> delete(boolean delete) {
...
+	 * Establish the session factory that will be used to create {@link Neo4jTemplate} instances
 	 * for interacting with Neo4j.
-	 * @param sessionFactory sessionFactory to be used.
+	 * @param neo4jTemplate neo4jTemplate to be used.
+	 * @return The current instance of the builder
+	 * @see Neo4jItemWriter#setNeo4jTemplate(Neo4jTemplate)
+	 */
+	public Neo4jItemWriterBuilder<T> neo4jTemplate(Neo4jTemplate neo4jTemplate) {
+		this.neo4jTemplate = neo4jTemplate;
+
```
Comment (dgray16) [inline spring-batch-neo4j/src/main/java/org/springframework/batch/extensions/neo4j/builder/Neo4jItemWriterBuilder.java:65]:
```suggestion

```

File: spring-batch-neo4j/src/test/java/org/springframework/batch/extensions/neo4j/Neo4jItemReaderTests.java [hunk 3]
```diff
@@ -16,69 +16,63 @@
...
-	private Iterable<String> result;
-	@Mock
-	private SessionFactory sessionFactory;
-	@Mock
-	private Session session;
+	@SuppressWarnings("unchecked")
+	@BeforeEach
+	void setup() {
+		neo4jTemplate = mock(Neo4jTemplate.class);
+		result = mock(List.class);
```
Comment (dgray16) [inline spring-batch-neo4j/src/test/java/org/springframework/batch/extensions/neo4j/Neo4jItemReaderTests.java:45]: I propose not to mock a `java.util.List`. It is very easy to create and it does not feel like external dependency which is usually mocked.

In the test you can write a simple statement like:

```
when(neo4jTemplate.findAll(...).thenReturn(List.of(...));
```

File: spring-batch-neo4j/src/test/java/org/springframework/batch/extensions/neo4j/Neo4jItemWriterTests.java [hunk 4]
```diff
@@ -16,134 +16,471 @@
...
 
-	@Test
-	public void testWriteItemsWithSession() throws Exception {
-		writer = new Neo4jItemWriter<>();
+		verify(this.neo4jDriver, new Times(2)).executableQuery("MATCH (MyEntity) WHERE MyEntity.idField = $id DETACH DELETE MyEntity");
+	}
 
-		writer.setSessionFactory(this.sessionFactory);
-		writer.afterPropertiesSet();
+	private static class MyEntity {
```
Comment (dgray16) [inline spring-batch-neo4j/src/test/java/org/springframework/batch/extensions/neo4j/Neo4jItemWriterTests.java:168]: Can we replace it with `Java Record`?

**Review by dgray16 — APPROVED** [review]
I did not have a chance to work with Neo4j so I cannot add a proper review for the logic, but dropping you some general ideas.

Also it would be great to have integration tests with [Testcotainers](https://java.testcontainers.org/modules/databases/neo4j/).

Anyway LGTM. Thanks for your efforts!

**Comment by meistermeier** [comment]
Thanks for your input, @dgray16 
I added the suggested changes. Don't have any preferences of whitespace formatting, so I adjusted it to spaces 😃 
Would be nice to get a last sanity check on this. I am happy to squash and merge it later.

**Review by dgray16 — APPROVED** [review]

**Pull request merged**